Index: backports-6.9.9/net/mac80211/rc80211_minstrel_ht.c
===================================================================
--- backports-6.9.9.orig/net/mac80211/rc80211_minstrel_ht.c
+++ backports-6.9.9/net/mac80211/rc80211_minstrel_ht.c
@@ -414,10 +414,11 @@ out:
  */
 static struct minstrel_rate_stats *
 minstrel_ht_ri_get_stats(struct minstrel_priv *mp, struct minstrel_ht_sta *mi,
-			  struct ieee80211_rate_status *rate_status)
+			 struct ieee80211_rate_status *rate_status)
 {
-	int group, idx;
+	int group, idx, rate_idx;
 	struct rate_info *rate = &rate_status->rate_idx;
+	struct ieee80211_supported_band *sband;
 
 	if (rate->flags & RATE_INFO_FLAGS_MCS) {
 		group = minstrel_ht_ri_get_group_idx(rate);
@@ -444,9 +445,12 @@ minstrel_ht_ri_get_stats(struct minstrel
 	}
 
 	group = MINSTREL_OFDM_GROUP;
-	for (idx = 0; idx < ARRAY_SIZE(mp->ofdm_rates[0]); idx++)
-		if (rate->legacy == minstrel_ofdm_bitrates[ mp->ofdm_rates[mi->band][idx] ])
+	sband = mp->hw->wiphy->bands[mi->band];
+	for (idx = 0; idx < ARRAY_SIZE(mp->ofdm_rates[0]); idx++) {
+		rate_idx = mp->ofdm_rates[mi->band][idx];
+		if (rate->legacy == sband->bitrates[rate_idx].bitrate)
 			goto out;
+	}
 
 	idx = 0;
 out:
@@ -1196,26 +1200,27 @@ minstrel_ht_ri_txstat_valid(struct minst
 			    struct minstrel_ht_sta *mi,
 			    struct ieee80211_rate_status *rate_status)
 {
-	int i;
+	int i, rate_idx;
+	struct rate_info *rate = &rate_status->rate_idx;
+	struct ieee80211_supported_band *sband;
 
 	if (!rate_status)
 		return false;
 	if (!rate_status->try_count)
 		return false;
 
-	if (rate_status->rate_idx.flags & RATE_INFO_FLAGS_MCS ||
-	    rate_status->rate_idx.flags & RATE_INFO_FLAGS_VHT_MCS)
+	if (rate->flags & RATE_INFO_FLAGS_MCS ||
+	    rate->flags & RATE_INFO_FLAGS_VHT_MCS)
 		return true;
 
-	for (i = 0; i < ARRAY_SIZE(mp->cck_rates); i++) {
-		if (rate_status->rate_idx.legacy ==
-		    minstrel_cck_bitrates[ mp->cck_rates[i] ])
+	for (i = 0; i < ARRAY_SIZE(mp->cck_rates); i++)
+		if (rate->legacy == minstrel_cck_bitrates[ mp->cck_rates[i] ])
 			return true;
-	}
 
-	for (i = 0; i < ARRAY_SIZE(mp->ofdm_rates); i++) {
-		if (rate_status->rate_idx.legacy ==
-		    minstrel_ofdm_bitrates[ mp->ofdm_rates[mi->band][i] ])
+	sband = mp->hw->wiphy->bands[mi->band];
+	for (i = 0; i < ARRAY_SIZE(mp->ofdm_rates[0]); i++) {
+		rate_idx = mp->ofdm_rates[mi->band][i];
+		if (rate->legacy == sband->bitrates[rate_idx].bitrate)
 			return true;
 	}
 
@@ -1920,7 +1925,7 @@ minstrel_ht_alloc(struct ieee80211_hw *h
 
 	minstrel_ht_init_cck_rates(mp);
 	for (i = 0; i < ARRAY_SIZE(mp->hw->wiphy->bands); i++)
-	    minstrel_ht_init_ofdm_rates(mp, i);
+		minstrel_ht_init_ofdm_rates(mp, i);
 
 	return mp;
 }
