--- a/net/mac80211/rc80211_minstrel_ht_api.c
+++ b/net/mac80211/rc80211_minstrel_ht_api.c
@@ -105,8 +105,8 @@ minstrel_ht_dump_sta(struct minstrel_pri
 	int i;
 
 	ofs += scnprintf(info + ofs, sizeof(info) - ofs, "%llx;sta;%s;%pM;%x;%x;",
-			(unsigned long long)ktime_get_boottime_ns(),
-			 type, mi->sta->addr, mi->overhead, mi->overhead_legacy);
+			(unsigned long long)ktime_get_real_fast_ns(),
+			type, mi->sta->addr, mi->overhead, mi->overhead_legacy);
 
 	ofs += scnprintf(info + ofs, sizeof(info) - ofs, "%x",
 			 mi->supported[0]);
@@ -450,8 +450,9 @@ void minstrel_ht_sta_remove(struct minst
 	if (!mp->monitor)
 		goto out;
 
+
 	ofs = scnprintf(info, sizeof(info), "%llx;sta;remove;%pM;;;\n",
-			(unsigned long long)ktime_get_boottime_ns(),
+			(unsigned long long)ktime_get_real_fast_ns(),
 			mi->sta->addr);
 	relay_write(mp->relay_ev, info, ofs);
 	relay_flush(mp->relay_ev);
@@ -473,8 +474,9 @@ void __minstrel_ht_report_tx_status(stru
 	if (!n_rates)
 		return;
 
+
 	ofs += scnprintf(txs, sizeof(txs), "%llx;txs;%pM;%x;%x;%x",
-			 (unsigned long long)ktime_get_boottime_ns(),
+			 (unsigned long long)ktime_get_real_fast_ns(),
 			 mi->sta->addr,
 			 info->status.ampdu_len,
 			 info->status.ampdu_ack_len,
@@ -504,9 +506,10 @@ void __minstrel_ht_report_rate_update(st
 	tp = minstrel_ht_get_tp_avg(mi, MI_RATE_GROUP(rate), MI_RATE_IDX(rate),
 				    mrs->prob_avg);
 
+
 	ofs = scnprintf(stat, sizeof(stat),
 			"%llx;stats;%pM;%x;%x;%x;%x;%x;%x;%x\n",
-			(unsigned long long)ktime_get_boottime_ns(),
+			(unsigned long long)ktime_get_real_fast_ns(),
 			mi->sta->addr, rate,
 			MINSTREL_TRUNC(mrs->prob_avg * 1000), tp,
 			mrs->last_success,
