--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -3719,10 +3719,15 @@ struct ieee80211_prep_tx_info {
  * @IEEE80211_FTRCTL_ADAPTIVE_SENSITIVITY: Adaptive sensitivity to
  * 	dynamically adjust the receive sensitivity by a certain algorithm.
  *
+ * @IEEE80211_FTRCTL_TPC: Transmit power control per packet/mrr stage. In
+ * 	case this is enabled (state == 1) the driver should accept setting
+ * 	transmit power to use per packet/per mrr stage (via ieee80211_sta_rates).
+ *
  * @NUM_IEEE80211_FTRCTL: Number of features.
  */
 enum ieee80211_feature_ctrl {
 	IEEE80211_FTRCTRL_ADAPTIVE_SENSITIVITY = 0,
+	IEEE80211_FTRCTRL_TPC,
 
 	/* keep last, obviously */
 	NUM_IEEE80211_FTRCTRL,
