Index: backports-6.9.9/include/net/mac80211.h
===================================================================
--- backports-6.9.9.orig/include/net/mac80211.h
+++ backports-6.9.9/include/net/mac80211.h
@@ -3826,18 +3826,27 @@ struct ieee80211_prep_tx_info {
  * and the driver ops get_feature_state and set_feature_state to dynamically
  * control the state of a particular feature at runtime.
  *
- * @IEEE80211_FTRCTL_ADAPTIVE_SENSITIVITY: Adaptive sensitivity to
+ * @IEEE80211_FTRCTRL_ADAPTIVE_SENSITIVITY: Adaptive sensitivity to
  * 	dynamically adjust the receive sensitivity by a certain algorithm.
  *
- * @IEEE80211_FTRCTL_TPC: Transmit power control per packet/mrr stage. In
+ * @IEEE80211_FTRCTRL_TPC: Transmit power control per packet/mrr stage. In
  * 	case this is enabled (state == 1) the driver should accept setting
  * 	transmit power to use per packet/per mrr stage (via ieee80211_sta_rates).
  *
+ * @IEEE80211_FTRCTRL_ACK_TPC: Control tx power of ACK frames.
+ * @IEEE80211_FTRCTRL_RTSCTS_TPC: Control tx power of RTS/CTS frames.
+ * @IEEE80211_FTRCTRL_CHIRP_TPC: Control tx power of CHIRP frames.
+ * @IEEE80211_FTRCTRL_RPT_TPC: Control tx power of RPT frames.
+ *
  * @NUM_IEEE80211_FTRCTL: Number of features.
  */
 enum ieee80211_feature_ctrl {
 	IEEE80211_FTRCTRL_ADAPTIVE_SENSITIVITY = 0,
 	IEEE80211_FTRCTRL_TPC,
+	IEEE80211_FTRCTRL_ACK_TPC,
+	IEEE80211_FTRCTRL_RTSCTS_TPC,
+	IEEE80211_FTRCTRL_CHIRP_TPC,
+	IEEE80211_FTRCTRL_RPT_TPC,
 
 	/* keep last, obviously */
 	NUM_IEEE80211_FTRCTRL,
