--- a/net/mac80211/rc80211_minstrel_ht_api.c
+++ b/net/mac80211/rc80211_minstrel_ht_api.c
@@ -12,7 +12,7 @@ extern u8 sample_table[SAMPLE_COLUMNS][M
 
 enum sta_cmd {
 	STA_CMD_PROBE,
-	STA_CMD_RATES,
+	STA_CMD_RC,
 };
 
 static void
@@ -77,7 +77,7 @@ minstrel_ht_read_api_info(struct seq_fil
 	seq_printf(s, "\n");
 
 	seq_printf(s, "#reset_stats;macaddr\n");
-	seq_printf(s, "#rates;macaddr;rates;counts\n");
+	seq_printf(s, "#rc;macaddr;rates;counts\n");
 	seq_printf(s, "#probe;macaddr;rate\n");
 
 	seq_printf(s, "#sample_table;cols;rows");
@@ -448,7 +448,7 @@ minstrel_ht_api_sta_cmd(struct minstrel_
 	case STA_CMD_PROBE:
 		ret = minstrel_ht_set_probe_rate(mi, args[1]);
 		break;
-	case STA_CMD_RATES:
+	case STA_CMD_RC:
 		ret = minstrel_ht_set_rates(mp, mi, args[1], args[2]);
 		break;
 	}
@@ -563,8 +563,8 @@ minstrel_ht_control_write(struct file *f
 		minstrel_ht_api_set_manual(mp, false);
 	else if (!strcmp(cur, "reset_stats"))
 		err = minstrel_ht_reset_stats(mp, pos);
-	else if (!strcmp(cur, "rates"))
-		err = minstrel_ht_api_sta_cmd(mp, STA_CMD_RATES, pos);
+	else if (!strcmp(cur, "rc") || !strcmp(cur, "rates"))
+		err = minstrel_ht_api_sta_cmd(mp, STA_CMD_RC, pos);
 	else if (!strcmp(cur, "probe"))
 		err = minstrel_ht_api_sta_cmd(mp, STA_CMD_PROBE, pos);
 	else
