--- a/net/mac80211/rc80211_minstrel_ht_api.c
+++ b/net/mac80211/rc80211_minstrel_ht_api.c
@@ -52,7 +52,7 @@ minstrel_ht_read_api_info(struct seq_fil
 	        seq_printf(s, ";airtime%d", i);
         seq_printf(s, "\n");
 
-	seq_printf(s, "#sta;action;macaddr;overhead_mcs;overhead_legacy");
+	seq_printf(s, "#sta;action;macaddr;iface;overhead_mcs;overhead_legacy");
 	for (i = 0; i < MINSTREL_GROUPS_NB; i++)
 	        seq_printf(s, ";mcs%d", i);
 	seq_printf(s, "\n");
@@ -191,13 +191,16 @@ static void
 minstrel_ht_dump_sta(struct minstrel_priv *mp, struct minstrel_ht_sta *mi,
 		     const char *type)
 {
+	struct sta_info *sta_info;
 	char info[64 + MINSTREL_GROUPS_NB * 4];
 	int ofs = 0;
 	int i;
 
-	ofs += scnprintf(info + ofs, sizeof(info) - ofs, "%llx;sta;%s;%pM;%x;%x;",
-			(unsigned long long)ktime_get_real_fast_ns(),
-			type, mi->sta->addr, mi->overhead, mi->overhead_legacy);
+	sta_info = container_of(mi->sta, struct sta_info, sta);
+	ofs += scnprintf(info + ofs, sizeof(info) - ofs, "%llx;sta;%s;%pM;%s;%x;%x;",
+			(unsigned long long)ktime_get_real_fast_ns(), type,
+			mi->sta->addr, sta_info->sdata->name, mi->overhead,
+			mi->overhead_legacy);
 
 	ofs += scnprintf(info + ofs, sizeof(info) - ofs, "%x",
 			 mi->supported[0]);
